### PR TITLE
Algorithm optimization

### PR DESCRIPTION
**Pull request by LevDenisov**
- Added test "test_segmentation_image.cpp ", which conducts segmentation of various images.

Changes in cell_grind.cpp:

- Added parallelization for multiple call of CellSegment constructors inside the CellGrid constructor;

- The order of function calls has been changed;

- A vector has been added to store Eigen::Map<const Eigen::Matrix<float, Eigen::Dynamic, 3, Eigen::RowMajor>> for more efficient data access

-----------------
22.07.2022:

Changes in plane_extractor.cpp

- Optimized the code in the PlaneExtractor::Impl::toImageLabels method

### CODE REVIEW

File: cpp/deplex/src/deplex/cell_grid.cpp [hunk 1]
```diff
@@ -18,6 +18,15 @@
 #include <utility>
 
 namespace deplex {
+
+inline void CellGrid::threadFunction(uint id_thread, config::Config const& config,
```
Comment (achains) [inline cpp/deplex/src/deplex/cell_grid.cpp]: What does "threadFunction" mean? Can we give it a clear name?

File: cpp/deplex/src/deplex/cell_grid.cpp [hunk 2]
```diff
@@ -33,11 +42,27 @@ CellGrid::CellGrid(Eigen::Matrix<float, Eigen::Dynamic, 3, Eigen::RowMajor> cons
 
   Eigen::Map<const Eigen::Matrix<float, Eigen::Dynamic, 3, Eigen::RowMajor>> cell_points(cell_continuous_points.data(),
                                                                                          cell_width_ * cell_height_, 3);
-#pragma omp parallel for default(none) shared(cell_continuous_points, config, cell_points)
+
+  std::vector<std::thread> threads;
+  std::vector<Eigen::Map<const Eigen::Matrix<float, Eigen::Dynamic, 3, Eigen::RowMajor>>> arr;
```
Comment (achains) [inline cpp/deplex/src/deplex/cell_grid.cpp]: What is arr?

File: cpp/deplex/src/deplex/cell_segment.h [hunk 3]
```diff
@@ -16,10 +16,13 @@
 #pragma once
 
 #include <cstdint>
+#include <thread>
 
 #include "cell_segment_stat.h"
 #include "deplex/config.h"
 
+#define sizeThreads 4 // Optimal 4 - 8
```
Comment (achains) [inline cpp/deplex/src/deplex/cell_segment.h]: Why do we need to define this variable?

Comment (achains) [inline cpp/deplex/src/deplex/cell_segment.h]: Snake case for variables

File: cpp/tests/test_segmentation_image.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+#include <deplex/plane_extractor.h>
+#include <deplex/utils/depth_image.h>
+#include <deplex/utils/eigen_io.h>
+
+TEST(ReadImage, count_image) {
+  const uint iterations = 500;
+  const std::string IMAGE = "icl_nuim";
+
+  std::__fs::filesystem::path data_dir = "/Users/denisovlev/Documents/CLionProjects/deplex/dataBenchmark";
```
Comment (achains) [inline cpp/tests/test_segmentation_image.cpp]: Hardcoded path...

File: cpp/tests/CMakeLists.txt [hunk 5]
```diff
@@ -29,7 +29,7 @@ add_executable(unit-tests
         test_plane_extractor.cpp
         test_config.cpp
         test_depth_image.cpp
-        )
+        test_segmentation_image.cpp)
```
Comment (achains) [inline cpp/tests/CMakeLists.txt]: I am quite unsure that we need this test. What does it check? It is more like an example, but we already have similar one.

**Review by achains — CHANGES_REQUESTED** [review]

File: cpp/deplex/src/deplex/cell_segment.h (hunk 3 above)
Comment (LevDenisov) [reply cpp/deplex/src/deplex/cell_segment.h]: This variable is necessary for calculating steps in parallel algorithms

File: cpp/deplex/src/deplex/cell_grid.cpp (hunk 1 above)
Comment (LevDenisov) [reply cpp/deplex/src/deplex/cell_grid.cpp]: Fixed

File: cpp/deplex/src/deplex/cell_grid.cpp (hunk 2 above)
Comment (LevDenisov) [reply cpp/deplex/src/deplex/cell_grid.cpp]: Fixed

File: cpp/deplex/src/deplex/cell_segment.h (hunk 3 above)
Comment (LevDenisov) [reply cpp/deplex/src/deplex/cell_segment.h]: Fixed

File: cpp/tests/test_segmentation_image.cpp (hunk 4 above)
Comment (LevDenisov) [reply cpp/tests/test_segmentation_image.cpp]: Will be deleted later

File: cpp/tests/CMakeLists.txt (hunk 5 above)
Comment (LevDenisov) [reply cpp/tests/CMakeLists.txt]: Will be deleted later

File: cpp/deplex/src/deplex/cell_segment.h [hunk 6]
```diff
@@ -21,7 +21,7 @@
 #include "cell_segment_stat.h"
 #include "deplex/config.h"
 
-#define sizeThreads 4 // Optimal 4 - 8
+#define size_threads 8 // Optimal 4 - 8
```
Comment (achains) [inline cpp/deplex/src/deplex/cell_segment.h]: Why now it is 8? What if I don't have 8 available threads

File: cpp/deplex/src/deplex/cell_segment.h (hunk 3 above)
Comment (achains) [reply cpp/deplex/src/deplex/cell_segment.h]: What I mean is why we need to use "define" directive. Can't we configure it dynamically or move to config file or configure it with CMake?

File: cpp/tests/test_segmentation_image.cpp (hunk 4 above)
Comment (achains) [reply cpp/tests/test_segmentation_image.cpp]: With this "local" test I can't see whether CI successfully passes all other tests. We'll have to delete it before moving further.

File: cpp/tests/CMakeLists.txt (hunk 5 above)
Comment (achains) [reply cpp/tests/CMakeLists.txt]: With this "local" test I can't see whether CI successfully passes all other tests. We'll have to delete it before moving further.

Comment (LevDenisov) [reply cpp/tests/CMakeLists.txt]: Text deleted

File: cpp/deplex/src/deplex/cell_segment.h (hunk 3 above)
Comment (LevDenisov) [reply cpp/deplex/src/deplex/cell_segment.h]: Moved

File: cpp/deplex/src/deplex/cell_segment.h (hunk 6 above)
Comment (LevDenisov) [reply cpp/deplex/src/deplex/cell_segment.h]: Moved

**Pull request closed without merging**